### PR TITLE
Add option to use linear scaling rule

### PR DESCRIPTION
**Pull request by michaelisc**
Add an option to use the linear scaling rule to automatically scale the learning rate when more or less than 8 gpus are used.

Results [box AP] on coco val2017:
1 GPU: 36.3
2 GPUs: 36.5
4 GPUs: 36.5
8 GPUs: 36.4

### CODE REVIEW

**Comment by hellock** [comment]
Thanks! Some minor linting errors are present. I will merge this PR once they are fixed.

**Comment by michaelisc** [comment]
Is there a way to find out, where linting fails? Fixing this by trial and error might work for this pull request but a lot harder for the other, more complex ones.

**Comment by hellock** [comment]
You can click the `Details` link below and then the job link, and all the errors are listed.

**Comment by michaelisc** [comment]
Looks like I had some pep8 irregularities. They should be fixed now.

**Comment by michaelisc** [comment]
I don't know if we should provide some evidence, that this works somewhere. I did some tests on coco and found 1, 2, 4 and 8 gpus to provide the same results when using the linear scaling rule.

**Comment by hellock** [comment]
It would be nice if you can attach the results of different gpus for reference, but I have not figured out a good place to add it. Maybe first in the description of this PR?

**Comment by michaelisc** [comment]
Here are the results:

1 GPU:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.363
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.580
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.392
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.217
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.401
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.465
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.303
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.487
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.512
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.332
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.551
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.643

2 GPUs:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.365
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.584
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.394
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.218
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.402
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.471
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.305
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.488
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.513
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.323
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.551
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.647

4 GPUs:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.365
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.582
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.394
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.215
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.400
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.466
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.306
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.489
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.513
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.328
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.551
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.645

8 GPUs:
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.582
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.391
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.213
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.402
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.464
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.304
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.487
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.511
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.322
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.552
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.641

**Comment by dmarnerides** [comment]
Hi both, 

I know this is a bit late, but coming back to this, is this applicable when the imgs_per_gpu setting changes? As I understand it, the linear scaling rule depends on the batch size and not the number of gpus directly.

**Comment by dmarnerides** [comment]
Hi all,

I know this is a bit late, but coming back to this issue, is this adjustment applicable to when imgs_per_gpu changes from 2 to say, 4 on a larger GPU? As I understand it, the linear scaling rule is dependent on the batch size, and not the number of gpus directly. 

Should that be accounted for when applying the LSR here?
e.g. 
```python
cfg.optimizer['lr'] = cfg.optimizer['lr'] * cfg.gpus* cfg.data.imgs_per_gpu / 16 
```

Also, the --gpus flag is applicable only for non-distributed training. When running the (default) distributed training with dist_train.sh, the number of gpus is set by --nproc_per_node here:

```bash
$PYTHON -m torch.distributed.launch --nproc_per_node=$GPUS \
    $(dirname "$0")/train.py $CONFIG --launcher pytorch ${@:3}
```
which maybe could also change to:
```bash
$PYTHON -m torch.distributed.launch --nproc_per_node=$GPUS \
    $(dirname "$0")/train.py $CONFIG --gpus $GPUS --launcher pytorch ${@:3}
```

What do you think?

Thanks

**Comment by michaelisc** [comment]
@dmarnerides you are right. It would be more elegant to use the batch size to scale the learning rate. I however did not do this because currently only the number of gpus can be selected through argparse and the number of images per gpu is selected in the config file. I am honestly not sure, what the most user friendly solution would be. The current solution has the advantage that it requires the user to specify the scaling at the point where he makes the change (number of images per gpu in the config file, number of gpus through argparse).

You are also right about distributed training. I never used distributed training, so I don't know what would be the best way to implement it there. Currently it is simply not available in distributed mode.
